File: codes/python/chap08/answer_A60.py
from collections import deque

# 입력
N = int(input())
A = list(map(int, input().split()))

# 스택 변화 재현
# (스택에는 (날짜, 주가) 튜플을 기록한다)
answer = [ None ] * N
level2 = deque()
for i in range(N):
	if i >= 1:
		level2.append((i, A[i - 1]))
		while len(level2) >= 1:
			kabuka = level2[-1][1] # 주가는 튜플의 2번째 요소
			if kabuka <= A[i]:
				level2.pop()
			else:
				break
	if len(level2) >= 1:
		answer[i] = level2[-1][0] # 날짜는 튜플의 1번째 요소
	else:
		answer[i] = -1

# answer를 공백으로 구분해서 출력
print(*answer)

File: codes/java/chap08/answer_A60.java
import java.util.*;
import java.io.*;

class Main {
	public static void main(String[] args) throws IOException {
		// 입력(고속의 입력을 위해, Scanner 대신 BufferedReader를 사용하고 있습니다)
		BufferedReader buff = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		st = new StringTokenizer(buff.readLine());
		int N = Integer.parseInt(st.nextToken());
		int[] A = new int[N + 1];
		st = new StringTokenizer(buff.readLine());
		for (int i = 1; i <= N; i++) {
			A[i] = Integer.parseInt(st.nextToken());
		}

		// 스택 변화 재현
		int[] answer = new int[N + 1];
		Stack<PairInt> level2 = new Stack<>();
		for (int i = 1; i <= N; i++) {
			if (i >= 2) {
				level2.push(new PairInt(i - 1, A[i - 1]));
				while (!level2.empty()) {
					int kabuka = level2.peek().second;
					if (kabuka <= A[i]) {
						level2.pop();
					}
					else {
						break;
					}
				}
			}
			// 기산일 특정
			if (!level2.empty()) {
				answer[i] = level2.peek().first;
			}
			else {
				answer[i] = -1;
			}
		}

		// 출력(고속 출력을 위해, System.out.println이 아닌 PrintWriter를 사용하고 있습니다)
		PrintWriter output = new PrintWriter(System.out);
		for (int i = 1; i <= N; i++) {
			if (i >= 2) {
				output.print(" ");
			}
			output.print(answer[i]);
		}
		output.println();
		output.flush();
	}
	
	// int 타입의 쌍의 클래스 PairInt
	static class PairInt {
		int first, second;
		public PairInt(int first, int second) {
			this.first = first;
			this.second = second;
		}
	}
}